grfmerge.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" Please adjust this date whenever revising the manpage.
.Dd August 05, 2010
.Dt GRFMerge 1
.Sh NAME
.Nm GRFMerge
.Nd A tool to apply a binary GRF diff, generated by GRFDiff, to a GRF.
.Sh SYNOPSIS
.Nm grfmerge
.Op Fl hly
.Ar GRD-file
.Op Ar GRF-file
.Pp
Change sprites in the GRF file to the new ones from the GRD file.
If the GRF file is not specified, GRFMerge will modify the one
which the GRD file was generated from.
.Sh OPTIONS
.Bl -tag
.It Fl h
Display the help message.
.It Fl l
Only show which sprites the GRD file contains, don't integrate them.
.It Fl y
Answer 'y' to all questions.
.El
.Sh SEE ALSO
The documentation in
.Pa /usr/share/doc/grfcodec/
.Sh AUTHOR
GRFMerge was written by Josef Drexler.
.Pp
This manual page was written by Remko Bijker.
.Sh COPYRIGHT
GRFMerge is Copyright (C) 2003 by Josef Drexler <jdrexler@uwo.ca>


Changelog
Changes in
- (devel) Detect UPX, fix some race issues in the Makefile, disable
  optimization if GCC 4.5.0 is detected (to work around an optimization bug),
  fix version.h generation if there is a version.h in /usr/include.
- (devel) Fix a number of compiler warnings.
- Clamp the arguments for -w and -h to 65535 as that is the maximum size of a
  PCX.
- Silence progress-meter output when stdout is not a console.
- Unify warnings and errors so they go to stderr and are prefixed by "Warning:"
  or "Error:".
- Allow hex escapes that have bit 31 set.
- After cropping a sprite, compress and write the correct number of bytes.
- Add -X to print PCX sprite numbers in hex
- (devel) Many compilation issues with different compilers and architectures.
- Fix bug in UTF-8 generation for escapes \U800..\U7FFF.
- Fix U+E0xx characters being output quoted.
- (devel) By Tron: code cleanup, make C functions and variables static where
  possible, remove dead and debugging code.
- Require two slashes before dismissing anything as a C++ style comment.
- Fix reading of real sprites with Unixy absolute paths.
- Check for pixels with palette index of 255 when encoding.
- Add -q to silence warnings.
- Add -c to automatically crop extraenous transparent blue pixels.
- Add escapes for the new action 2 operations 0E, 0F, 11, 12, 13, 14, 15, 16
- Prevent quotes and escapes in sprite 0.
- Add escapes for long-format intro dates.

Changes in 0.9.10 (r819)
- (bugfix) Correct reading of \7c, \7C, \D| and \D&. (DaleStan)

Changes in 0.9.9 (r653)
- By Patchman:
  - Check for sprites that extend beyond the end of the PCX file.
  - Update versioning system to show last-comitted revision for non-release
    builds.
- By DaleStan:
  - (devel) Add release targets to the Makefiles
  - Check the palette of the PCX file before encoding from it.
    (Override a failed check with -f)
  - Introduce several backslash escape sequences; see the wiki for further
    information.
    (disable with -x or -xx)
  - Do not apply colormap to sprites that like character glyphs.
    (Apply colormap to all sprites with -M instead of -m)
  - Enable reading of GIMP palette files with -p.
  - (devel) Terse build mode; get the verbose mode with make V=1.
  - (devel) Touch version.h if and only if it changes.

Changes in 0.9.8 (r197)
- By DaleStan:
  - (devel) grfcodec (but not grfdiff or grfmerge) can now be built in
    MS Visual Studio 2003.
  - Slightly change the plaintext quoting algorithm.
  - (devel) Change several memcpy()s to pointer assignments.
  - (devel) Plug several memory leaks
  - (devel) Source code and Makefile changes to grfdiff and grfmerge build
    with both Makefiles, on both Cygwin and Linux.
  - Make code gcc 4 -Wall clean.
  - Use NFORenum's NFO reading engine:
    - Sprite numbers are now optional, and need not be correct if present.
    - Sprite lengths need not be correct. Set to 0 to supress the incorrect
      length warnings.
    - Real sprites are checked for format compliance.
  - Make 0-byte pseudo sprites an error.
  - (bugfix) Prevent binary includes from including directories on OSes that
    would othewise allow it.
  - (bugfix) It is now possible for a real sprite to start above the
    preceeding sprite (the infamous decreasing ypos bug)
  - grfcodec now exits with 0 on success.

Changes in 0.9.7
- support including binary files using "  4 ** data/dingdong.wav" syntax
- by Oskar: support // and # comments within pseudo-sprites
- fixed incorrect error message while encoding if .nfo file could not be read
  (now it reports "Cannot read %s" instead of Creating sprites/: File exists")
- made code -Wall clean
- remove myalloc stuff

Changes in 0.9.6
- by pasky:
  * made it possible to compile grfcodec on Linux
  * made compatible with gcc 3.x
  * when decoding, text strings are written in plain text (disable with -t)
  * supports plain text in .nfo files
  * fixed bug when first encoding a grf file wouldn't rename .new to .grf

Changes in 0.9.5
- self-extracting executables now work on Win2K etc., which doesn't pass
  the full path to the C library
- GRFMerge will just skip .grf files it can't find
- making sprites larger doesn't work if they're at the right edge; complains
  that ofsx too large; instead just use background colour
- GRFDiff doesn't like grfdiff -l 1254 ... (single value in -l), complains
  that there were no differences and makes no output
- allow GRFMerge to read GRFDiff's .exe files, both for -l and other options

Changes in 0.9.4
- fixed yet another crash while compressing
- sprites no longer need correct numbers in the .NFO file; just use "-1" for
  each number and it will be assumed to be correct

Changes in 0.9.3
- fixed crash due to buffer overrun while compressing
- slightly improved compression efficiency and speed
- GRFCodec now handles spaces in file/directory names properly
- GRFCodec now supports comments in .NFO files (from //, ; or # to end of line)
- speed up PCX processing

Changes in 0.9.2
- fixed crash if pcx file not found
- fixed crash if nfo file has no version info header

Changes in 0.9.1
- bug fix by Marcin: wrong order of test in encode function

Changes in release April 20 2002
- -p now actually uses the parameter you give it

Changes in release Dec 03 2001
- made -p option a valid option
- improve compression speed even more using assembly optimizations
- compression is now on by default, disable with -c-
- fixed bug where encoding would fail if the grf file didn't exist already
- fixed bug where decoding would fail creating the sprites directory
- corrected some mistakes in grf.txt (checksum, and initial offsets for tiles)

Changes in release March 26 2001
- fix bug that didn't flush last row if it only contained a single sprite
- fix bug with no height given (would make PCX have height -1)
- show percentage done while decoding
- fix compression algorithm
- improve compression performance
- try without compression if data is inconsistent after decompressing our

src/grfmerge.c

/*****************************************\
*                                         *
* GRFMerge - A program to integrate a     *
*            .GRD file generated by       *
*            GRFDiff into the respective  *
*            GRF file.                    *
*                                         *
*                                         *
* Copyright (C) 2003 by Josef Drexler     *
*                      <jdrexler@uwo.ca>  *
*                                         *
* Permission granted to copy and redist-  *
* ribute under the terms of the GNU GPL.  *
* For more info please read the file      *
* COPYING which should have come with     *
* this file.                              *
*                                         *
\*****************************************/

#include <stdlib.h>
#include <stdio.h>
#include <getopt.h>
#include <stdarg.h>
#include <string.h>
#include <ctype.h>
#include <errno.h>

#include "typesize.h"
#include "version.h"
#include "grfcomm.h"

static const U32 GRDmagic = 0x67fb49ad;
#define TEMPFILE "grfmerge.tmp"

static int alwaysyes = 0;
static char *grdfile = NULL;
static long grdofs = 0;
static int onlyshow = 0, issfx = 0;

static void usage(void)
{
	printf(
		"\nUsage:\n"
		"    GRFMerge [options] %s[<GRF-File>]\n"
		"	Change sprites in the GRF file to the new ones from the GRD file.\n"
		"	If the GRF file is not specified, GRFMerge will modify the one\n"
		"	which the GRD file was generated from.\n"
		"\n"
		"Options:\n"
		"	-h  Show this help\n"
		"	-l  Only show which sprites the GRD file contains, don't integrate them\n"
		"	-y  Answer 'y' to all questions\n"
		"\n"
		"GRFMerge is Copyright (C) 2003 by Josef Drexler\n"
		"It may be freely copied and distributed.\n",
		issfx?"":"<GRD-File> "
		);

	exit(1);
}

static int checkisselfextr(const char *exe)
{
	static const char *action = "reading exe";
	FILE *f;
	char c[3];
	U8 r, e;
	U32 magic = 0;

	f = fopen(exe, "rb");
	if (!f && errno == ENOENT) {
		// try appending .exe for Win2k
		char *altexe = (char*) malloc(strlen(exe)+5);
		strcpy(altexe, exe);
		strcat(altexe, ".exe");
		f = fopen(altexe, "rb");
		free(altexe);
	}
	if (!f) return 0;

	cfread(action, c, 2, 1, f);
	if (c[0] != 'M' || c[1] != 'Z')
		return 0;

	fseek(f, 0x1c, SEEK_SET);
	cfread(action, c, 2, 1, f);
	c[2] = 0;

	if (!strcmp(c, "JD")) {
		cfread(action, &r, 1, 1, f);
		cfread(action, &e, 1, 1, f);

		grdofs = (long) r * (1L << e);
		fseek(f, grdofs, SEEK_SET);

		cfread(action, &magic, 4, 1, f);
	}
	fclose(f);

	return issfx = magic == GRDmagic;
}

static void die(const char *text, ...)
{
	va_list argptr;

	va_start(argptr, text);
	vprintf(text, argptr);
	va_end(argptr);

	exit(2);
}

static int yesno(const char *txt)
{
	char c;

	printf(" [Y/N] ");
	if (alwaysyes) {
		printf("Y\n");
	} else {
		c = tolower(getc(stdin));

		// skip the newline
		getc(stdin);

		if (c != 'y') {
			printf("%s", txt);
			return 0;
		}
	}
	return 1;
}

static char *block = NULL;
#define BLOCKSIZE 8192
static void copyblock(size_t size, FILE *from, FILE *to)
{
	size_t thisblock;

	while (size > 0) {
		if (size > BLOCKSIZE)
			thisblock = BLOCKSIZE;
		else
			thisblock = size;

		cfread("copying block", block, 1, thisblock, from);

		if (to) cfwrite("copying block", block, 1, thisblock, to);

		size -= thisblock;
	}
}

// this function reads a piece of data, and copies it to the other files
static void copy(void *data, size_t size, size_t n, FILE *from, FILE *to)
{
	size_t res;

	res = fread(data, size, n, from);
	if (res != n)
		die("Error while reading, wanted %d, got %d: %s", n, res, strerror(errno));

	if (to) {
		res = fwrite(data, size, n, to);
		if (res != n)
			die("Error while writing, wanted %d, wrote %d: %s", n, res, strerror(errno));
	}
}

//
// Copy the data of a sprite from one file to another
// this is complicated because for some sprites, the GRF file stores
// the uncompressed length instead of how many bytes are in the file
//
// This code mostly copied from sprites.c, but to keep this program
// small I didn't link in sprites.c and removed the unnecessary steps
// from decodesprite (we're not actually interested in the uncompressed data,
// just the length)
//

static U16 copysprite(FILE *from, FILE *to)
{
	U8 info, ofs;
	S8 code;
	U16 size;
	size_t reallen;

#ifdef DEBUG
	if (from)
		printf("Copying from pos. %ld\n", ftell(from));
	if (to)
		printf("Copying to pos. %ld\n", ftell(to));
#endif

	if (feof(from)) return 0;

	copy(&size, 2, 1, from, to);

	if (!size) return 0;	// EOF is also indicated by a zero-size sprite

	copy(&info, 1, 1, from, to);

	if (info == 0xff) {	// verbatim data
		copyblock(size, from, to);
		return 1;
	}

	if (info & 2) {	// size is compressed size
		copyblock(size-1, from, to);	// already copied one byte
		return 1;
	}

	// the size given is the uncompressed size, so we need to go through
	// the uncompression routine until we have enough bytes
	copyblock(7, from, to);
	size -= 8;
	while (size > 0) {
		copy(&code, 1, 1, from, to);

		if (code < 0) {
			copy(&ofs, 1, 1, from, to);
			reallen = -(code >> 3);
		} else {
			reallen = code ? code : 128;
			copyblock(reallen, from, to);
		}
		if (size < reallen)
			die("\nOops, got too many bytes. How did that happen?\n"
			"Size is %d, len is %d at GRF file pos %ld\n",
			size, reallen, ftell(from));

		size -= reallen;
	}
	return 1;
}

static void skipsprite(FILE *f)
{
	copysprite(f, NULL);
}

static void showpct(long now, long total, int spriteno, int *pct)
{
	int newpct = 100L*now/total;

	if (newpct == *pct)
		return;

	printf("\rSprite%5d  Done:%3d%%  \r", spriteno, newpct);
	*pct = newpct;
}

static int mergeset(FILE *grd, const char *grffile)
{
	static const char *action = "reading GRD";
	FILE *grf = NULL, *tmp = NULL;
	char grflen, *grfname;
	const char *c;
	char tempfile[16];
	U16 version, i, j, numsprites, spriteno, curno;
	int lastfrom = -2, lastto = lastfrom, lastpct = -1;
	long grfsize = 0;
	U32 dummy;
	int skipped = onlyshow;

	cfread(action, &version, 2, 1, grd);

	if (version > 1) die("This is a GRD file version %d, I don't know how to handle that.\n", version);

	cfread(action, &numsprites, 2, 1, grd);
	cfread(action, &grflen, 1, 1, grd);

	grfname = (char*) malloc(grflen + 4);		// +4 for .bak extension (safety margin)
	if (!grfname) die("Out of memory.\n");

	cfread(action, grfname, 1, grflen, grd);

	if (onlyshow) {
		printf("Generated from: %s.grf\nSprites in file: ", grfname);
	} else {
		if (grffile) {
			c=strrchr(grffile, '\\');
			if (!c) c=strchr(grffile, ':');
			if (!c)
				c=grffile;
			else
				c++;

			if (strrchr(c, '.'))
				j=strrchr(c, '.') - c;
			else
				j=strlen(c);

			if (strnicmp(c, grfname, j)) {
				printf("Warning, this GRD file was generated from %s.GRF.\n", grfname);
				printf("Are you sure you want to apply it to %s?", grffile);
				if (!yesno("Skipping file\n")) {
					for (i=0; i<numsprites; i++) {
						cfread(action, &spriteno, 2, 1, grd);
						skipsprite(grd);
					}
					return 1;
				};
			}
		} else {
			strcat(grfname, ".grf");
			grffile = grfname;
		}

		grf = fopen(grffile, "rb");
		if (!grf) {
			printf("Can't open %s: %s. File skipped.\n", grffile, strerror(errno));
			skipped = 1;
		}
	}
	if (!skipped) {
		fseek(grf, 0, SEEK_END);
		grfsize = ftell(grf);
		fseek(grf, 0, SEEK_SET);

		c=strchr(grffile, ':');
		if (c) {
			strncpy(tempfile, grffile, sizeof(tempfile)-1);
			i=c-grffile+1;
			if (i<sizeof(tempfile))
				tempfile[i]=0;
		} else
			tempfile[0] = 0;
		strncat(tempfile, TEMPFILE, sizeof(tempfile)-strlen(tempfile)-1);

		tmp = fopen(tempfile, "wb");
		if (!tmp) die("Can't open %s: %s\n", tempfile, strerror(errno));

		printf("Writing temporary file %s\n", tempfile);
	}

	curno = 0;
	for (i=0; i<numsprites; i++) {
		cfread(action, &spriteno, 2, 1, grd);

		if (skipped) {
			if (onlyshow) {
				if (spriteno != lastto+1) {
					if (lastfrom != lastto) printf("-%d", lastto);
					if (lastfrom >= 0) printf(", ");
					lastfrom = spriteno;
					printf("%d", lastfrom);
				}
				lastto = spriteno;
			}

			skipsprite(grd);
		} else {

			while (++curno <= spriteno) {
				copysprite(grf, tmp);
				showpct(ftell(grf), grfsize, curno, &lastpct);
			}

			skipsprite(grf);
			copysprite(grd, tmp);
			showpct(ftell(grf), grfsize, curno, &lastpct);
		}
	}

	if (onlyshow) {
		if (lastfrom < 0)
			printf("No sprites.");
		else if (lastfrom != lastto)
			printf("-%d", lastto);
		printf("\n");
	} else if (!skipped) {
		// copy remaining sprites, if any
		for (; copysprite(grf, tmp); )
			showpct(ftell(grf), grfsize, ++curno, &lastpct);
		showpct(grfsize, grfsize, curno, &lastpct);

		// write the dummy checksum
		dummy = 0;
		cfwrite("writing dummy checksum", &dummy, 4, 1, tmp);
	}

	if (tmp) fclose(tmp);
	if (grf) fclose(grf);

	if (!skipped) {
		char* c;

		printf("\nDone\n");

		// rename grf to bak if bak doesn't exist
		strcpy(block, grffile);
		c = strrchr(block, '.');
		if (!c) c = block + strlen(block);
		strcpy(c, ".bak");

		tmp = fopen(block, "rb");

		if (!tmp && (errno == ENOENT)) {
			// .bak doesn't exist, rename .grf to .bak
			printf("Renaming %s to %s\n", grffile, block);
			if (rename(grffile, block)) {
				printf("Error while renaming: %s\n", strerror(errno));
				printf("Shall I delete it instead?");
				if (!yesno("Aborted.\n")) {
					die("");
				}
				errno = EEXIST;		// go delete it
			} else {
				errno = ENOENT;		// don't try to delete it
			}
		}
		if (tmp || (errno != ENOENT)) {
			printf("Deleting %s\n", grffile);
			if (remove(grffile))
				die("Error while deleting: %s\n", strerror(errno));
		}

		// rename tmp to grf
		printf("Renaming %s to %s\n", tempfile, grffile);
		if (rename(tempfile, grffile))
			die("Error while renaming: %s\n", strerror(errno));

		printf("All done!\n");
	}

	free(grfname);

	return 1;
}

static int domerge(int argc, char **argv)
{
	FILE *grd;
	U32 magic;
	int first = 1;

	grd = fopen(grdfile, "rb");
	if (!grd) die("Can't open %s: %s\n", grdfile, strerror(errno));

	block = (char*) malloc(BLOCKSIZE);
	if (!block) die("Out of memory.\n");

	// read all sets of sprites
	while (1) {
		if (feof(grd))
			break;

		fseek(grd, grdofs, SEEK_SET);
		cfread("reading GRD", &magic, 4, 1, grd);

		if (magic != GRDmagic) {
			if (first)
				die("This is not a GRD file.\n");

			break;
		}

		first = 0;

		mergeset(grd, optind < argc ? argv[optind++] : NULL);

		grdofs = ftell(grd);

	};

	if (grd) fclose(grd);

	free(block);

	return 1;
}

int main(int argc, char **argv)
{
	char opt;

	printf("GRFMerge version " GRFCODECVER " - Copyright (C) 2003 by Josef Drexler\n");

#ifdef WIN32
	//  debugint();
#endif

	checkisselfextr(argv[0]);

	while (1) {
		opt = getopt(argc, argv, "hlyv");

		if (opt == (char) EOF)
			break;

		switch (opt) {
		case 'l':
			onlyshow = 1;
			break;
		case 'v':
			/* The version is already printed. */
			return 0;
		case 'y':
			alwaysyes = 1;
			break;
		default:
			usage();
		}
	}
	if (!issfx && (optind < argc) )	// no first arg if we're self-extracting
		grdfile = argv[optind++];

	if (grdfile)
		// see if the GRD file specified is a self-extracting executable
		// also sets grdofs to skip the .exe code
		checkisselfextr(grdfile);
	else {
		// no GRD file specified; see if we're a self-extracting executable
		if (issfx)
			grdfile = argv[0];
		else {
			printf("No GRD file specified!\n");
			exit(2);
		}
	}

	return domerge(argc, argv);
}
